### ISSUE TITLE
GPU Training Errors with NaN Loss in Ultralytics YOLO

### ISSUE THREAD

**Issue by lavrentijav** [ultralytics/ultralytics#17812]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug
During training on the GPU, I encounter NaN values in the loss, which prevents the model from training effectively. The training log shows NaN values, while training on the CPU does not produce these errors.

the dataset on which the AI model was trained: [dataset.zip](https://github.com/user-attachments/files/17921372/dataset.zip)
result trained AI model: [trained_model.zip](https://github.com/user-attachments/files/17921355/trained_model.zip)

LOGS:
```
C:\Users\Public\Projects\Python\YOLO\venv\Scripts\python.exe C:\Users\Public\Projects\Python\YOLO\train.py 
Load model
Start training
Ultralytics 8.3.38 🚀 Python-3.11.9 torch-2.5.1+cu118 CUDA:0 (NVIDIA GeForce GTX 1660, 6144MiB)
engine\trainer: task=detect, mode=train, model=yolo11n.pt, data=C:\Users\Public\Projects\Python\YOLO\Datasets\tiny\data.yaml, epochs=20, time=None, patience=100, batch=8, imgsz=434, save=True, save_period=-1, cache=False, device=0, workers=8, project=None, name=yolo11n_test11, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=None, format=torchscript, keras=False, optimize=False, int8=False
[trained_model.zip](https://github.com/user-attachments/files/17921355/trained_model.zip)
, dynamic=False, simplify=True, opset=None, workspace=None, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, bgr=0.0, mosaic=1.0, mixup=0.0, copy_paste=0.0, copy_paste_mode=flip, auto_augment=randaugment, erasing=0.4, crop_fraction=1.0, cfg=None, tracker=botsort.yaml, save_dir=runs\detect\yolo11n_test11
Overriding model.yaml nc=80 with nc=1

                   from  n    params  module                                       arguments                     
  0                  -1  1       464  ultralytics.nn.modules.conv.Conv             [3, 16, 3, 2]                 
  1                  -1  1      4672  ultralytics.nn.modules.conv.Conv             [16, 32, 3, 2]                
  2                  -1  1      6640  ultralytics.nn.modules.block.C3k2            [32, 64, 1, False, 0.25]      
  3                  -1  1     36992  ultralytics.nn.modules.conv.Conv             [64, 64, 3, 2]                
  4                  -1  1     26080  ultralytics.nn.modules.block.C3k2            [64, 128, 1, False, 0.25]     
  5                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]              
  6                  -1  1     87040  ultralytics.nn.modules.block.C3k2            [128, 128, 1, True]           
  7                  -1  1    295424  ultralytics.nn.modules.conv.Conv             [128, 256, 3, 2]              
  8                  -1  1    346112  ultralytics.nn.modules.block.C3k2            [256, 256, 1, True]           
  9                  -1  1    164608  ultralytics.nn.modules.block.SPPF            [256, 256, 5]                 
 10                  -1  1    249728  ultralytics.nn.modules.block.C2PSA           [256, 256, 1]                 
 11                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 12             [-1, 6]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 13                  -1  1    111296  ultralytics.nn.modules.block.C3k2            [384, 128, 1, False]          
 14                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 15             [-1, 4]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 16                  -1  1     32096  ultralytics.nn.modules.block.C3k2            [256, 64, 1, False]           
 17                  -1  1     36992  ultralytics.nn.modules.conv.Conv             [64, 64, 3, 2]                
 18            [-1, 13]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 19                  -1  1     86720  ultralytics.nn.modules.block.C3k2            [192, 128, 1, False]          
 20                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]              
 21            [-1, 10]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 22                  -1  1    378880  ultralytics.nn.modules.block.C3k2            [384, 256, 1, True]           
 23        [16, 19, 22]  1    430867  ultralytics.nn.modules.head.Detect           [1, [64, 128, 256]]           
YOLO11n summary: 319 layers, 2,590,035 parameters, 2,590,019 gradients, 6.4 GFLOPs

Transferred 448/499 items from pretrained weights
ClearML Task: created new task id=06d13e08ee7046ccb0353cb79b4fea7d
2024-11-26 17:51:42,584 - clearml.Task - INFO - No repository found, storing script code instead
ClearML results page: https://app.clear.ml/projects/5c84b1ae4b8641f8a679628532cd4bd0/experiments/06d13e08ee7046ccb0353cb79b4fea7d/output/log
ClearML Initialized a new task. If you want to run remotely, please add clearml-init and connect your arguments before initializing YOLO.
Freezing layer 'model.23.dfl.conv.weight'
AMP: running Automatic Mixed Precision (AMP) checks...
AMP: checks passed ✅
WARNING ⚠️ imgsz=[434] must be multiple of max stride 32, updating to [448]
train: Scanning C:\Users\Public\Projects\Python\YOLO\Datasets\tiny\train\labels.cache... 172 images, 92 backgrounds, 0 corrupt: 100%|██████████| 172/172 [00:00<?, ?it/s]
val: Scanning C:\Users\Public\Projects\Python\YOLO\Datasets\tiny\valid\labels.cache... 51 images, 24 backgrounds, 0 corrupt: 100%|██████████| 51/51 [00:00<?, ?it/s]
Plotting labels to runs\detect\yolo11n_test11\labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.002, momentum=0.9) with parameter groups 81 weight(decay=0.0), 88 weight(decay=0.0005), 87 bias(decay=0.0)
Image sizes 448 train, 448 val
Using 8 dataloader workers
Logging results to runs\detect\yolo11n_test11
Starting training for 20 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       1/20     0.633G        nan        nan        nan          8        448: 100%|██████████| 22/22 [00:06<00:00,  3.63it/s]
2024-11-26 17:54:10,354 - clearml - INFO - NaN value encountered. Reporting it as '0.0'. Use clearml.Logger.set_reporting_nan_value to assign another value
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:00<00:00,  7.08it/s]
                   all         51         27          0          0          0          0
  0%|          | 0/22 [00:00<?, ?it/s]
      Epoch    GPU_mem   box_loss   cls_loss  

 dfl_loss  Instances       Size
       2/20     0.652G        nan        nan        nan          1        448: 100%|██████████| 22/22 [00:05<00:00,  4.35it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:00<00:00,  8.70it/s]
                   all         51         27          0          0          0          0
  0%|          | 0/22 [00:00<?, ?it/s]
      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       3/20     0.654G        nan        nan        nan          2        448: 100%|██████████| 22/22 [00:05<00:00,  4.16it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:00<00:00,  7.84it/s]
                   all         51         27          0          0          0          0
  0%|          | 0/22 [00:00<?, ?it/s]
      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       4/20     0.654G        nan        nan        nan          5        448: 100%|██████████| 22/22 [00:04<00:00,  4.52it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:00<00:00,  9.03it/s]
                   all         51         27          0          0          0          0
  0%|          | 0/22 [00:00<?, ?it/s]
```
#then it's the same

### Environment

```
C:\Users\Public\Projects\Python\YOLO\venv\Scripts\python.exe C:\Users\Public\Projects\Python\YOLO\test_code.py 
Ultralytics 8.3.38 🚀 Python-3.11.9 torch-2.5.1+cu118 CUDA:0 (NVIDIA GeForce GTX 1660, 6144MiB)
Setup complete ✅ (8 CPUs, 23.9 GB RAM, 496.4/926.2 GB disk)

OS                  Windows-10-10.0.19045-SP0
Environment         Windows
Python              3.11.9
Install             pip
RAM                 23.95 GB
Disk                496.4/926.2 GB
CPU                 Intel Xeon E3-1225 V2 3.20GHz
CPU count           8
GPU                 NVIDIA GeForce GTX 1660, 6144MiB
GPU count           1
CUDA                11.8

numpy               ✅ 2.1.3>=1.23.0
numpy               ✅ 2.1.3<2.0.0; sys_platform == "darwin"
matplotlib          ✅ 3.9.2>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 11.0.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.14.1>=1.4.1
torch               ✅ 2.5.1+cu118>=1.8.0
torch               ✅ 2.5.1+cu118!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.20.1+cu118>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 6.1.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.12>=2.0.0

Process finished with exit code 0
```

```
PS C:\Users\Lavrenitijav> nvidia-smi
Tue Nov 26 17:13:43 2024
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 560.94                 Driver Version: 560.94         CUDA Version: 12.6     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                  Driver-Model | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce GTX 1660      WDDM  |   00000000:01:00.0  On |                  N/A |
| 46%   39C    P0             15W /  130W |     453MiB /   6144MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+

+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A      1760    C+G   ...siveControlPanel\SystemSettings.exe      N/A      |
|    0   N/A  N/A      5572    C+G   ...\cef\cef.win7x64\steamwebhelper.exe      N/A      |
|    0   N/A  N/A      8936    C+G   ...\PyCharm 2024.2.4\bin\pycharm64.exe      N/A      |
|    0   N/A  N/A     10076    C+G   C:\Windows\explorer.exe                     N/A      |
|    0   N/A  N/A     10536    C+G   ....Search_cw5n1h2txyewy\SearchApp.exe      N/A      |
|    0   N/A  N/A     12808    C+G   D:\Prog\Telegram Desktop\Telegram.exe       N/A      |
|    0   N/A  N/A     13468    C+G   ...t.LockApp_cw5n1h2txyewy\LockApp.exe      N/A      |
|    0   N/A  N/A     13736    C+G   ...oogle\Chrome\Application\chrome.exe      N/A      |
|    0   N/A  N/A     13860    C+G   ...17.0_x64__vfvw9svesycw6\Y.Music.exe      N/A      |
|    0   N/A  N/A     14096    C+G   ...CBS_cw5n1h2txyewy\TextInputHost.exe      N/A      |
|    0   N/A  N/A     14568    C+G   ...oogle\Chrome\Application\chrome.exe      N/A      |
+-----------------------------------------------------------------------------------------+
```

### Minimal Reproducible Example
from ultralytics import YOLO

if __name__ == '__main__':
    print("Load model")
    # Load the model
    model = YOLO('yolo11n.pt') 

    print("Start training")
    # Train the model
    results = model.train(
        data="C:\\Users\\Public\\Projects\\Python\\YOLO\\Datasets\\tiny\\data.yaml",
        epochs=20,
        imgsz=434,  # Ensure the image size is appropriate
        batch=8,
        name='yolo11n_test',
        device="0",  # Specify GPU device
    )

    print("Training is finished")
    print("Model exported to", model.export())

### Additional

Dear Ultralytics Team,

I am experiencing an issue with training a YOLO model using the Ultralytics library on a GPU. While training on the CPU, I do not encounter any NaN values, but when I switch to the GPU, the training fails with NaN losses. I would appreciate your assistance in resolving this issue.

Error: During training on the GPU, I encounter NaN values in the loss, which prevents the model from training effectively. The training log shows NaN values, while training on the CPU does not produce these errors.

Context:

I have tried various configurations and adjustments to the training parameters, including learning rate adjustments and batch size changes, but the issue persists.
The dataset is correctly formatted, and I have verified that the annotations are accurate.
I am using the latest version of the Ultralytics library.
Request for Assistance: I kindly request your guidance on how to resolve the NaN loss issue when training on the GPU. Are there specific parameters or configurations that I should adjust to prevent this from occurring? Any insights or recommendations would be greatly appreciated.

Thank you for your support!

Best regards,
lavrentijav

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @lavrentijav, thank you for bringing this up and for your detailed report! 🚀

We recommend reviewing the [Ultralytics Docs](https://docs.ultralytics.com) for helpful guidance, especially [Training Tips](https://docs.ultralytics.com/guides/model-training-tips/) to achieve the best results during your experiments.

It seems you're encountering a 🐛 NaN loss issue during GPU training. To assist you better, here are some steps you could try and share with us:

1. **Data Check**: Ensure your dataset annotations are properly formatted, and all the data (images and labels) adhere to the YOLO format. You can use the `ultralytics.utils.checks.yaml_check()` function to verify your YAML file or share specific information about your dataset.
2. **Image Size**: Your training logs display a warning about `imgsz`. Please ensure it's divisible by 32. You might want to try `imgsz=416` or `imgsz=640` (standard YOLO sizes).
3. **Minimum Reproducible Example (MRE)**: While you've shared a code snippet (thank you!), it's essential to ensure the MRE is self-contained:
    - If possible, include dummy example images and annotations that can provoke the issue.
    - Simplify the code to the smallest version that still reproduces the issue, possibly excluding external integrations like ClearML for initial testing.
4. **Environment Update**: Please ensure you’re using the latest version of the `ultralytics` package and all dependencies:
   ```bash
   pip install -U ultralytics
   ```
5. **Further Debugging**:
   - Try running a quick sanity check on CPU and GPU by setting `device='cpu'` or `device='0'` respectively.
   - Experiment with disabling AMP (`amp=False`) to check if mixed-precision training is causing the instability.

## Upgrade

Please ensure you're running the latest Ultralytics YOLO version with all minimum requirements in a [**Python>=3.8**](https://www.python.org/) environment:
```bash
pip install -U ultralytics
```

## Collaboration

Feel free to check out these community spaces for additional support or brainstorming:
- [Discord](https://ultralytics.com/discord) 🎧 for real-time discussions
- [Discourse](https://community.ultralytics.com) for detailed help threads
- [Subreddit](https://reddit.com/r/ultralytics) for sharing insights and troubleshooting tips

Lastly, if CI tests verify it’s not an issue within the framework itself, you’ll find the CI status badge here:  
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

An Ultralytics engineer will assist you shortly—thank you for your patience! 😊

**Comment by Y-T-G** [comment]
Use `amp=False`

**Comment by Y-T-G** [comment]
Or `batch=3`

**Comment by Y-T-G** [comment]
https://www.reddit.com/r/Ultralytics/comments/1eqfl7t/psa_gpus_that_have_issues_with_amp_training/

**Comment by VdLMV** [comment]
Suggestion try different image size. On most gpus the image size (both height and width) must be a multiple of 32, for cpus this restriction is in general more relaxed.

**Comment by lavrentijav** [comment]
> Use `amp=False`

This is not working

**Comment by glenn-jocher** [comment]
Thank you for the suggestion! As the logs indicate, the image size was automatically adjusted to 448 (a multiple of 32) from 434, so this requirement is already being met. However, experimenting with larger image sizes could still be worth exploring to investigate any potential impact.
